### PR TITLE
Hooks accumulators

### PR DESCRIPTION
**Pull request by bartekgorny**
This PR addresses #980 and is only for review and discussion

Proposed changes include:
* change to ejabberd_hooks:run so that it passes an empty map as first arg
* modification to all handlers that are called by :run 
* modification of 20 folded hooks so that they accept and return a map
* a new module for encapsulating accumulators

It is freshly rebased and passes most tests.

### CODE REVIEW

File: apps/ejabberd/src/mod_last.erl [hunk 1]
```diff
@@ -242,14 +245,21 @@ get_last_info(LUser, LServer) ->
...
 
--spec remove_user(ejabberd:user(), ejabberd:server()) -> ok | {error, term()}.
-remove_user(User, Server) ->
+%% #rh
+-spec remove_user(map(), ejabberd:user(), ejabberd:server()) -> map() | {error, term()}.
+remove_user(Acc, User, Server) ->
     LUser = jid:nodeprep(User),
     LServer = jid:nameprep(Server),
-    ?BACKEND:remove_user(LUser, LServer).
+    case ?BACKEND:remove_user(LUser, LServer) of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_last.erl]: According to **Elvis**:
> Don't use macros (like BACKEND on line 253) as module names.

File: apps/ejabberd/src/mod_amp.erl [hunk 2]
```diff
@@ -56,20 +56,31 @@ check_packet(Packet = #xmlel{attrs = Attrs}, Event) ->
 
 -spec check_packet(#xmlel{}, jid(), amp_event()) -> #xmlel{} | drop.
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_amp.erl]: According to **Elvis**:
> The spec in line 57 uses a record, please define a type for the record and use that instead.

File: apps/ejabberd/src/mod_disco.erl [hunk 3]
```diff
@@ -419,28 +429,29 @@ process_sm_iq_info(From, To, #iq{type = Type, lang = Lang, sub_el = SubEl} = IQ)
...
-           [#xmlel{name = <<"identity">>,
-                   attrs = [{<<"category">>, <<"account">>},
-  {<<"type">>, <<"registered">>}]}];
-       _ ->
-           []
-           end.
-
-
--spec get_sm_features(empty | any(),
+get_sm_identity(#{sm_identity := Ids} = Acc, _From, #jid{luser = LUser, lserver=LServer}, _Node, _Lang) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Elvis**:
> Line 437 is too long: get_sm_identity(#{sm_identity := Ids} = Acc, _From, #jid{luser = LUser, lserver=LServer}, _Node, _Lang) ->.

File: apps/ejabberd/src/mod_disco.erl [hunk 4]
```diff
@@ -196,41 +198,48 @@ process_local_iq_info(From, To, #iq{type = Type, lang = Lang,
...
 get_local_identity(Acc, _From, _To, Node, _Lang) when is_binary(Node) ->
     Acc.
 
 
--spec get_local_features(Acc :: 'empty' | {'error',_} | {'result',_},
+-spec get_local_features(Acc :: map(),% | {'error',_} | {'result',_},
                         From :: ejabberd:jid(),
                         To :: ejabberd:jid(),
                         Node :: binary(),
                         Lang :: ejabberd:lang()) -> {'error',_} | {'result',_}.
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Elvis**:
> Missing space right "," on line 221

File: apps/ejabberd/src/mod_bosh.erl [hunk 5]
```diff
@@ -135,8 +135,9 @@ stop(_Host) ->
 %% Hooks handlers
 %%--------------------------------------------------------------------
 
-node_cleanup(Node) ->
-    ?BOSH_BACKEND:node_cleanup(Node).
+node_cleanup(Acc, Node) ->
+    Res = ?BOSH_BACKEND:node_cleanup(Node),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_bosh.erl]: According to **Elvis**:
> Remove the dynamic function call on line 139. Only modules that define callbacks should make dynamic calls.

File: apps/ejabberd/src/ejabberd_c2s.erl [hunk 6]
```diff
@@ -1175,9 +1179,9 @@ handle_info({send_filtered, Feature, From, To, Packet}, StateName, StateData) ->
     end,
     fsm_next_state(StateName, StateData);
 handle_info({broadcast, Type, From, Packet}, StateName, StateData) ->
-    Recipients = ejabberd_hooks:run_fold(
+    #{recipients := Recipients} = ejabberd_hooks:run_fold(
 		   c2s_broadcast_recipients, StateData#state.server,
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_c2s.erl]: According to **Elvis**:
> Line 1183 has a tab at column 0.

File: apps/ejabberd/src/mod_disco.erl [hunk 7]
```diff
@@ -255,26 +264,27 @@ domain_to_xml(Domain) ->
     #xmlel{name = <<"item">>, attrs = [{<<"jid">>, Domain}]}.
 
 
--spec get_local_services(Acc :: 'empty' | {'error',_} | {'result',_},
+-spec get_local_services(Acc :: map(),
                          From :: ejabberd:jid(),
                          To :: ejabberd:jid(),
                          Node :: binary(),
                          Lang :: ejabberd:lang()) -> {'error',_} | {'result',_}.
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Elvis**:
> Missing space right "," on line 271

File: apps/ejabberd/src/ejabberd_c2s.erl [hunk 8]
```diff
@@ -2868,7 +2875,7 @@ maybe_resume_session(NextState, El, StateData) ->
 do_resume_session(SMID, El, [{_, Pid}], StateData) ->
     try
         {ok, OldState} = ?GEN_FSM:sync_send_event(Pid, resume),
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_c2s.erl]: According to **Elvis**:
> Don't use macros (like GEN_FSM on line 2877) as module names.

File: apps/ejabberd/src/mod_disco.erl (hunk 7 above)
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Elvis**:
> Missing space right "," on line 271

File: apps/ejabberd/src/mod_caps.erl [hunk 9]
```diff
@@ -144,11 +144,12 @@ read_caps([_ | Tail], Result) ->
...
-user_send_packet(#jid{luser = User, lserver = Server} = From,
+user_send_packet(Acc,
+                 #jid{luser = User, lserver = Server} = From,
                  #jid{luser = User, lserver = Server,
                       lresource = <<"">>},
                  #xmlel{name = <<"presence">>, attrs = Attrs,
-                        children = Els} = Pkt) ->
+                        children = Els}) ->
     Type = xml:get_attr_s(<<"type">>, Attrs),
     if Type == <<"">>; Type == <<"available">> ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_caps.erl]: According to **Elvis**:
> Replace the 'if' expression on line 154 with a 'case' expression or function clauses.

File: apps/ejabberd/src/ejabberd_c2s.erl [hunk 10]
```diff
@@ -1175,9 +1179,9 @@ handle_info({send_filtered, Feature, From, To, Packet}, StateName, StateData) ->
     end,
     fsm_next_state(StateName, StateData);
 handle_info({broadcast, Type, From, Packet}, StateName, StateData) ->
-    Recipients = ejabberd_hooks:run_fold(
+    #{recipients := Recipients} = ejabberd_hooks:run_fold(
 		   c2s_broadcast_recipients, StateData#state.server,
-		   [],
+		   #{recipients => []},
 		   [StateData#state.server, StateData, Type, From, Packet]),
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_c2s.erl]: According to **Elvis**:
> Line 1185 has a tab at column 0.

File: apps/ejabberd/src/mod_disco.erl [hunk 11]
```diff
@@ -169,17 +169,19 @@ process_local_iq_info(From, To, #iq{type = Type, lang = Lang,
...
+            Res = ejabberd_hooks:run_fold(disco_local_features,
                                          Host,
-                                         empty,
-                                         [From, To, Node, Lang]) of
-                {result, Features} ->
+                                         A3,
+                                         [From, To, Node, Lang]),
+            case mongoose_perdix:to_map(Res) of
+                #{features := Features, info := Info, local_identity := Identity} ->
                     ANode = case Node of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Elvis**:
> The expression on line 185 and column 29 is nested beyond the maximum level of 3.

File: apps/ejabberd/src/mod_disco.erl [hunk 12]
```diff
@@ -196,41 +198,48 @@ process_local_iq_info(From, To, #iq{type = Type, lang = Lang,
...
-                            {<<"name">>, <<"ejabberd">>}]}];
+                            {<<"name">>, <<"ejabberd">>}]}],
+    mongoose_perdix:append(local_identity, NIds, Acc).
+
 get_local_identity(Acc, _From, _To, Node, _Lang) when is_binary(Node) ->
     Acc.
 
 
--spec get_local_features(Acc :: 'empty' | {'error',_} | {'result',_},
+-spec get_local_features(Acc :: map(),% | {'error',_} | {'result',_},
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Elvis**:
> Missing space right "," on line 217

File: apps/ejabberd/src/mod_bosh.erl (hunk 5 above)
Comment (gadgetci) [inline apps/ejabberd/src/mod_bosh.erl]: According to **Elvis**:
> Don't use macros (like BOSH_BACKEND on line 139) as module names.

File: apps/ejabberd/src/mod_disco.erl (hunk 4 above)
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Elvis**:
> Missing space right "," on line 221

File: apps/ejabberd/src/mod_last.erl [hunk 13]
```diff
@@ -223,11 +223,14 @@ get_last(LUser, LServer) ->
 count_active_users(LServer, Timestamp) ->
     ?BACKEND:count_active_users(LServer, Timestamp).
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_last.erl]: According to **Elvis**:
> Don't use macros (like BACKEND on line 224) as module names.

File: apps/ejabberd/src/mod_amp.erl [hunk 14]
```diff
@@ -97,7 +108,7 @@ amp_features() ->
 -spec process_amp_rules(#xmlel{}, jid(), amp_event(), amp_rules()) -> #xmlel{} | drop.
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_amp.erl]: According to **Elvis**:
> The spec in line 108 uses a record, please define a type for the record and use that instead.

File: apps/ejabberd/src/ejabberd_c2s.erl [hunk 15]
```diff
@@ -1155,10 +1159,10 @@ handle_info({force_update_presence, LUser}, StateName,
     end,
     {next_state, StateName, NewStateData};
 handle_info({send_filtered, Feature, From, To, Packet}, StateName, StateData) ->
-    Drop = ejabberd_hooks:run_fold(c2s_filter_packet, StateData#state.server,
-				   true, [StateData#state.server, StateData,
+    Res = ejabberd_hooks:run_fold(c2s_filter_packet, StateData#state.server,
+				   #{drop => true}, [StateData#state.server, StateData,
 					  Feature, To, Packet]),
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_c2s.erl]: According to **Elvis**:
> Line 1164 has a tab at column 0.

File: apps/ejabberd/src/ejabberd_c2s.erl [hunk 16]
```diff
@@ -1175,9 +1179,9 @@ handle_info({send_filtered, Feature, From, To, Packet}, StateName, StateData) ->
     end,
     fsm_next_state(StateName, StateData);
 handle_info({broadcast, Type, From, Packet}, StateName, StateData) ->
-    Recipients = ejabberd_hooks:run_fold(
+    #{recipients := Recipients} = ejabberd_hooks:run_fold(
 		   c2s_broadcast_recipients, StateData#state.server,
-		   [],
+		   #{recipients => []},
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_c2s.erl]: According to **Elvis**:
> Line 1184 has a tab at column 0.

File: apps/ejabberd/src/mod_amp.erl [hunk 17]
```diff
@@ -111,13 +122,15 @@ process_amp_rules(Packet, From, Event, Rules) ->
...
 
 -spec determine_strategy(#xmlel{}, jid(), amp_event()) -> amp_strategy().
 determine_strategy(Packet, From, Event) ->
     To = message_target(Packet),
     ejabberd_hooks:run_fold(amp_determine_strategy, host(From),
-                            amp_strategy:null_strategy(), [From, To, Packet, Event]).
+                            #{strategy => amp_strategy:null_strategy()},
+                            [From, To, Packet, Event]).
 
 -spec fold_apply_rules(#xmlel{}, jid(), amp_strategy(), [amp_rule()]) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_amp.erl]: According to **Elvis**:
> The spec in line 135 uses a record, please define a type for the record and use that instead.

File: apps/ejabberd/src/mod_http_notification.erl [hunk 18]
```diff
@@ -31,15 +31,16 @@ stop(Host) ->
     ejabberd_hooks:delete(user_send_packet, Host, ?MODULE, on_user_send_packet, 100),
     ok.
 
-on_user_send_packet(From, To, Packet) ->
+on_user_send_packet(Acc, From, To, Packet) ->
     Body = exml_query:path(Packet, [{element, <<"body">>}, cdata], <<>>),
     Mod = get_callback_module(),
     case Mod:should_make_req(Packet, From, To) of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_http_notification.erl]: According to **Elvis**:
> Remove the dynamic function call on line 37. Only modules that define callbacks should make dynamic calls.

File: apps/ejabberd/src/ejabberd_hooks.erl [hunk 19]
```diff
@@ -228,23 +229,25 @@ code_change(_OldVsn, State, _Extra) ->
 %%% Internal functions
 %%%----------------------------------------------------------------------
 
-run1([], _Hook, _Args) ->
-    ok;
-run1([{_Seq, Module, Function} | Ls], Hook, Args) ->
+run1([], _Hook, Acc, _Args) ->
+    Acc;
+run1([{_Seq, Module, Function} | Ls], Hook, Acc, Args) ->
     Res = if is_function(Function) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_hooks.erl]: According to **Elvis**:
> Replace the 'if' expression on line 235 with a 'case' expression or function clauses.

File: apps/ejabberd/src/ejabberd_c2s.erl [hunk 20]
```diff
@@ -1155,10 +1159,10 @@ handle_info({force_update_presence, LUser}, StateName,
     end,
     {next_state, StateName, NewStateData};
 handle_info({send_filtered, Feature, From, To, Packet}, StateName, StateData) ->
-    Drop = ejabberd_hooks:run_fold(c2s_filter_packet, StateData#state.server,
-				   true, [StateData#state.server, StateData,
+    Res = ejabberd_hooks:run_fold(c2s_filter_packet, StateData#state.server,
+				   #{drop => true}, [StateData#state.server, StateData,
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_c2s.erl]: According to **Elvis**:
> Line 1163 has a tab at column 0.

File: apps/ejabberd/src/mod_disco.erl [hunk 21]
```diff
@@ -392,15 +402,15 @@ process_sm_iq_info(From, To, #iq{type = Type, lang = Lang, sub_el = SubEl} = IQ)
                 true ->
                     Host = To#jid.lserver,
                     Node = xml:get_tag_attr_s(<<"node">>, SubEl),
-                    Identity = ejabberd_hooks:run_fold(disco_sm_identity,
+                    #{sm_identity := Identity} = ejabberd_hooks:run_fold(disco_sm_identity,
                                                        Host,
-                                                       [],
+                                                       #{sm_identity => []},
                                                        [From, To, Node, Lang]),
                     case ejabberd_hooks:run_fold(disco_sm_features,
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Elvis**:
> The expression on line 409 and column 21 is nested beyond the maximum level of 3.

File: apps/ejabberd/src/mod_disco.erl [hunk 22]
```diff
@@ -468,21 +479,22 @@ get_user_resources(User, Server) ->
...
                      Mod
              end,
     Serverinfo_fields = get_fields_xml(Host, Module),
-    [#xmlel{name = <<"x">>,
+    Info = [#xmlel{name = <<"x">>,
             attrs = [{<<"xmlns">>, ?NS_XDATA}, {<<"type">>, <<"result">>}],
             children = [#xmlel{name = <<"field">>,
                                attrs = [{<<"var">>, <<"FORM_TYPE">>}, {<<"type">>, <<"hidden">>}],
                                children = [#xmlel{name = <<"value">>,
                                                   children = [#xmlcdata{content = ?NS_SERVERINFO}]}]}]
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Elvis**:
> Line 495 is too long:                                                   children = [#xmlcdata{content = ?NS_SERVERINFO}]}]}].

File: apps/ejabberd/src/mod_caps.erl [hunk 23]
```diff
@@ -275,31 +279,32 @@ c2s_filter_packet(InAcc, Host, C2SState, {pep_message, Feature}, To, _Packet) ->
...
+c2s_broadcast_recipients(#{recipients := Rec} = Acc, Host, C2SState,
                          {pep_message, Feature}, _From, _Packet) ->
-    case ejabberd_c2s:get_aux_field(caps_resources,
+    NRec = case ejabberd_c2s:get_aux_field(caps_resources,
                                     C2SState)
     of
         {ok, Rs} ->
-            gb_trees_fold(fun (USR, Caps, Acc) ->
+            gb_trees_fold(fun (USR, Caps, Ac) ->
                                   case lists:member(Feature,
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_caps.erl]: According to **Elvis**:
> The expression on line 297 and column 35 is nested beyond the maximum level of 3.

File: apps/ejabberd/src/mod_disco.erl [hunk 24]
```diff
@@ -142,9 +142,9 @@ process_local_iq_items(From, To, #iq{type = Type, lang = Lang, sub_el = SubEl} =
 
             case ejabberd_hooks:run_fold(disco_local_items,
                                          Host,
-                                         empty,
+                                         #{local_items => []},
                                          [From, To, Node, Lang]) of
-                {result, Items} ->
+                #{local_items := Items} ->
                     ANode = case Node of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Elvis**:
> The expression on line 148 and column 29 is nested beyond the maximum level of 3.

File: apps/ejabberd/src/ejabberd_c2s.erl [hunk 25]
```diff
@@ -2345,7 +2349,10 @@ fsm_reply(Reply, StateName, StateData) ->
 is_ip_blacklisted(undefined) ->
     false;
 is_ip_blacklisted({IP,_Port}) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_c2s.erl]: According to **Elvis**:
> Missing space right "," on line 2351

File: apps/ejabberd/src/mod_caps.erl [hunk 26]
```diff
@@ -158,13 +159,13 @@ user_send_packet(#jid{luser = User, lserver = Server} = From,
...
+    Acc.
 
-user_receive_packet(#jid{lserver = Server}, From, _To,
+user_receive_packet(Acc, #jid{lserver = Server}, From, _To,
                     #xmlel{name = <<"presence">>, attrs = Attrs,
-                           children = Els} = Pkt) ->
+                           children = Els}) ->
     Type = xml:get_attr_s(<<"type">>, Attrs),
     IsRemote = not lists:member(From#jid.lserver, ?MYHOSTS),
     if IsRemote and
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_caps.erl]: According to **Elvis**:
> Replace the 'if' expression on line 171 with a 'case' expression or function clauses.

File: apps/ejabberd/src/ejabberd_c2s.erl (hunk 8 above)
Comment (gadgetci) [inline apps/ejabberd/src/ejabberd_c2s.erl]: According to **Elvis**:
> Remove the dynamic function call on line 2877. Only modules that define callbacks should make dynamic calls.

File: apps/ejabberd/src/mod_amp.erl [hunk 27]
```diff
@@ -111,13 +122,15 @@ process_amp_rules(Packet, From, Event, Rules) ->
 %% @doc ejabberd_hooks helpers
 -spec verify_support(binary(), amp_rules()) -> [amp_rule_support()].
 verify_support(Host, Rules) ->
-    ejabberd_hooks:run_fold(amp_verify_support, Host, [], [Rules]).
+    Res = ejabberd_hooks:run_fold(amp_verify_support, Host, #{supported => []}, [Rules]),
+    maps:get(supported, Res).
 
 -spec determine_strategy(#xmlel{}, jid(), amp_event()) -> amp_strategy().
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_amp.erl]: According to **Elvis**:
> The spec in line 128 uses a record, please define a type for the record and use that instead.

File: apps/ejabberd/src/mod_disco.erl [hunk 28]
```diff
@@ -196,41 +198,48 @@ process_local_iq_info(From, To, #iq{type = Type, lang = Lang,
...
 get_local_identity(Acc, _From, _To, <<>>, _Lang) ->
-    Acc ++ [#xmlel{name = <<"identity">>,
+    NIds = [#xmlel{name = <<"identity">>,
                    attrs = [{<<"category">>, <<"server">>},
                             {<<"type">>, <<"im">>},
-                            {<<"name">>, <<"ejabberd">>}]}];
+                            {<<"name">>, <<"ejabberd">>}]}],
+    mongoose_perdix:append(local_identity, NIds, Acc).
+
 get_local_identity(Acc, _From, _To, Node, _Lang) when is_binary(Node) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_disco.erl]: According to **Compiler**:
> function get_local_identity/5 already defined

**Comment by gadgetci** [comment]
**compiler** failed with exit status: 1

**Comment by michalwski** [comment]
I assume this is still work in progress but please remember to rebase, sqash and reword some commits.

**Pull request closed without merging**